Add localizedTime() validation rule

ypnos-web opened the pull request:
Functionality described in #8323 

Test is still missing. The test would go along the lines of creating I18n::Time and I18n::Date objects of various dates and using their string output as input to the rule next to invalid strings.

However I first ask for feedback regarding how to write proper tests.

ypnos-web committed:
localizedTime()

Date and/or time string validation that uses `I18n::Time` to parse the date. This means parsing is locale dependent.

ypnos-web committed:
localizedTime()

Proxy method for new Validation::localizedTime()

dereuromark commented on src/Validation/Validation.php:
```diff
@@ -457,6 +458,33 @@ public static function time($check)
+        if (empty($methods[$type])) {
+            throw new \InvalidArgumentException('Unsupported parser type given.');
+        }
+        return (!is_null(Time::{$methods[$type]}($check, $format)));
```
is_null should be `!== null`

dereuromark commented on src/Validation/Validation.php:
```diff
@@ -457,6 +458,33 @@ public static function time($check)
+     *
+     * @param string|\DateTime $check a date string or object (will always pass)
+     * @param string $type Parser type, one out of 'date', 'time', and 'datetime'
+     * @param string|int $format any format accepted by IntlDateFormatter
```
|null missing

markstory commented on src/Validation/Validation.php:
```diff
@@ -457,6 +458,33 @@ public static function time($check)
+     */
+    public static function localizedTime($check, $type = 'datetime', $format = null)
+    {
+        if ($check instanceof DateTime) {
```
You might want to check `DateTimeInterface` instead to cover DateTimeImmutable.

ypnos-web replied:
The same check is present in the original date/time methods, might be worth revisiting them as well.

markstory commented:
Code looks good. For tests I would add the following cases, for each of Date/Time/DateTime:

* Invalid value
* Valid value
* Valid but incorrectly formatted date based on the locale.
* Empty string

lorenzo commented:
ping @ypnos-web

ypnos-web commented:
Hello, I am currently on vacation, I plan to finish work (tests) on this PR in a week.

ypnos committed:
Merge remote-tracking branch 'upstream/master' into patch-4

ypnos committed:
improve code style, doc

ypnos committed:
Check DateTimeInterface which also covers DateTimeImmutable

ypnos committed:
improve doc

ypnos committed:
Make Validator::localizedTime() fit testing framework and add test

ypnos committed:
Add test for Validation::localizedTime()

Please note that PHP's IntlDateFormatter date parsing is quite lenient.
For whatever reason, it would parse 31-12-2006 as 07-12-2008 (basically
adding up the 31 months on 0-12-2006) instead of rejecting the date.

ypnos-web commented:
I am finished with tests and suggest to merge.

dereuromark commented:
I like it!

markstory commented:
Looks like [there is a test failure](https://travis-ci.org/cakephp/cakephp/jobs/116112884). Other than failing tests this is looking good.

ypnos committed:
remove ill-advised test

As with dates, PHP also parses nonsensical time values.
12.90am becomes 1.30am.

ypnos committed:
Add localizedTime() to DateTime object test

ypnos committed:
Use classes must be in alphabetical order

ypnos-web commented:
This is the test that `12.90am` should not be a time (but PHP does create a time object out of it). It seems I got confused when I tested the tests.

Fixed.
